### ISSUE TITLE
(yaml ffi-help-rt) is outdated

### ISSUE THREAD

**Issue by rekado** [mwette/guile-libyaml#3]
I wrote a Guix package definition for this library and intend to add it to Guix:

```scheme
(define-public guile-libyaml
  (let ((commit "f5d33a6880e96571d3cb079ed7755ffc156cac46")
        (revision "1"))
    (package
      (name "guile-libyaml")
      (version (git-version "0" revision commit))
      (source
       (origin
         (method git-fetch)
         (uri (git-reference
               (url "https://github.com/mwette/guile-libyaml")
               (commit commit)))
         (file-name (git-file-name name version))
         (sha256
          (base32
           "12x91983fh1j39zy7kbk19acc1rqdh8515ddx1mh7l26j04k9wgq"))))
      (build-system gnu-build-system)
      (arguments
       `(#:modules (((guix build guile-build-system)
                     #:prefix guile:)
                    ,@%gnu-build-system-modules)
         #:imported-modules ((guix build guile-build-system)
                             ,@%gnu-build-system-modules)
         #:tests? #false ; there are none
         #:phases
         (modify-phases %standard-phases
           (delete 'configure)
           (add-after 'unpack 'remove-unused-files
             (lambda* (#:key inputs #:allow-other-keys)
               (for-each delete-file
                         '("guix.scm" "demo1.yml" "demo1.scm"
                           "yaml/libyaml.scm"
                           ;; This file is mismatched with the generated FFI code.
                           "yaml/ffi-help-rt.scm"))
               (copy-file (string-append (assoc-ref inputs "nyacc")
                                         "/share/guile/site/3.0/system/ffi-help-rt.scm")
                          "yaml/ffi-help-rt.scm")
               (substitute* "yaml/ffi-help-rt.scm"
                 (("system ffi-help-rt") "yaml ffi-help-rt"))
               #true))
           (add-before 'build 'build-ffi
             (lambda* (#:key inputs #:allow-other-keys)
               (invoke "guild" "compile-ffi"
                       "--no-exec" ; allow us to patch the generated file
                       "yaml/libyaml.ffi")
               (substitute* "yaml/libyaml.scm"
                 (("system ffi-help-rt") "yaml ffi-help-rt")
                 (("dynamic-link \"libyaml\"")
                  (format #false "dynamic-link \"~a/lib/libyaml\""
                          (assoc-ref inputs "libyaml"))))
               #true))
           (replace 'build
             (assoc-ref guile:%standard-phases 'build))
           (delete 'install))))
      (inputs
       `(("guile" ,guile-3.0)
         ("libyaml" ,libyaml)))
      (propagated-inputs
       `(("guile-bytestructures" ,guile-bytestructures)))
      (native-inputs
       `(("nyacc" ,nyacc)))
      (home-page "https://github.com/mwette/guile-libyaml")
      (synopsis "Guile wrapper for libyaml")
      (description
       "This package provides a simple yaml module for Guile using the
ffi-helper from nyacc.")
      (license license:lgpl3+))))
```

I had to replace `(yaml ffi-help-rt)` with an up-to-date copy of `(system ffi-help-rt)`, because the version of `nyacc` I'm using to compile the FFI file generates code that is not included in the old copy of `(yaml ffi-help-rt)`.

I also had to patch the generated Scheme code to refer to `(yaml ffi-help-rt)` --- it looks like the generated file in this repository was modified.  It would perhaps be a good idea to add this copying and patching to a Makefile.

**Comment by mwette** [comment]
This package was written to not need the nyacc distribution.  Are you suggesting it be made to depend on nyacc?
What is the minimum guix supported build paradigm?  Just a makefile?

**Comment by mwette** [comment]
And if you use emacs, how to you get the quasi-quoted list ``(#:modules ...) to indent like that?

**Comment by rekado** [comment]
> This package was written to not need the nyacc distribution. Are you suggesting it be made to depend on nyacc?

I understand that.  But since `libyaml.scm` file is not source code --- it is generated from `libyaml.ffi` --- I need to rebuild it from source for inclusion in Guix.  And that's where the mismatch originates: I'm using the current version of nyacc, which generates code that is available in `(system ffi-help-rt)`, but not in the old copy at `(yaml ffi-help-rt)`.  So in addition to rebuilding the `.ffi` file to `.scm` I also need to patch it to preserve the intent of making `guile-libyaml` not need `nyacc` at runtime.

I wonder if it would make sense to add this mode of operation to compile-ffi, to make it also copy and rename the `(system ffi-help-rt)` module for inclusion in the generated library, e.g. when "--standalone" is provided.

>  What is the minimum guix supported build paradigm? Just a makefile?

There's no minimum.  We can add the package as I pasted it above.  It would be nice if there was a release tag, though, because we prefer not to add arbitrary commits.  For the sake of inclusion in other distributions perhaps it would be nice to have a Makefile, though.

**Comment by rekado** [comment]
> And if you use emacs, how to you get the quasi-quoted list ``(#:modules ...) to indent like that?

The Guix repository comes with a `.dir-locals.el`, which sets up a few indentation rules, but I also use [Emacs Guix](https://git.savannah.gnu.org/cgit/guix/emacs-guix.git/).

**Comment by mwette** [comment]
Doing the "--standalone" option turns out to be a bit complex (i.e., searching for the file).
I'm going to just leave ffi-help-rt module as a reference from the nyacc distribution.

**Comment by mwette** [comment]
rekado, can you update the guix file wrt this commit: 6316099dd0500d49cd2f0defe23bc0eacb05fad0
Once I get that I will merge to master and tag with a version number.

**Comment by arunisaac** [comment]
Hi,

I ran into some issues when trying out the 6316099 commit from the `devel` branch. I built `yaml/libyaml.scm` using `guild compile-ffi yaml/libyaml.ffi`, and ran `./demo1.scm` but got the following backtrace.

```
Backtrace:
In ice-9/boot-9.scm:
  1752:10  9 (with-exception-handler _ _ #:unwind? _ #:unwind-for-type _)
In unknown file:
           8 (apply-smob/0 #<thunk 7fe6e66e3600>)
In ice-9/boot-9.scm:
    724:2  7 (call-with-prompt _ _ #<procedure default-prompt-handler (k proc)>)
In ice-9/eval.scm:
    619:8  6 (_ #(#(#<directory (guile-user) 7fe6e66e7c80>)))
In ice-9/boot-9.scm:
   2835:4  5 (save-module-excursion _)
  4380:12  4 (_)
In demo1.scm:
     4:12  3 (_)
In yaml.scm:
   120:17  2 (read-yaml-file _)
    95:39  1 (cnvt-node #<bytestructure 0x7fe6e31759c0>)
In /gnu/store/mbkn78jm1mzf67lwdcr9gpj7sf2bj3ji-profile/share/guile/site/3.0/system/ffi-help-rt.scm:
    459:8  0 (unwrapper _ _ _ _)

/gnu/store/mbkn78jm1mzf67lwdcr9gpj7sf2bj3ji-profile/share/guile/site/3.0/system/ffi-help-rt.scm:459:8: In procedure unwrapper:
Tried to follow void pointer.
```

I am using nyacc 1.04.0, libyaml 0.2.5 and guile 3.0.7. Any help much appreciated!

Thanks in advance! :-)

**Comment by mwette** [comment]
I was able to re-create this error.

**Comment by mwette** [comment]
the above crash issue will be fixed with 1.05.1.

**Comment by arunisaac** [comment]
> the above crash issue will be fixed with 1.05.1.

Thank you! I'll check again once it's out.

**Comment by mwette** [comment]
I believe all resolved in commit 2bdacb72a65ab63264b2edc9dac9692df7ec9b3e and with nyacc-1.05.1 just released.

**Comment by arunisaac** [comment]
It works. Thank you! I have also updated the Guix nyacc and guile-libyaml packages.
